### PR TITLE
Add a triple tiling expert for gemm for experiments.

### PR DESCRIPTION
**Pull request by hanhanW**
This PR creates a new expert which contains three level of tiling -- distribution, L1, register. In the PR, the L1 tiling level only considers reduction loop. I haven't observed perf improvements when tiling parallel dims for L1 level, so it's not added at this moment. We need more investigation about it.

The PR improves miniLM bert model 5%. All the improvements happens on large reduction dim cases.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
> Missing a bit of context. What is this trying to achieve? What does `expt` mean?
> 
> It general, it looks good. My only concern is that we are adding yet another independent pipeline for matmul. I'm worried about the scalability/composition of this. Since you are introducing a "benefit" function, I wonder if we could integrate this new strategy as part of one of the main pipelines, similar to what we did for peeling and use that strategy based on the benefit function. Does that make any sense?

Second this. I think we need to (a) consolidate different experiments into smaller number of pipelines, and maybe discard things that are not relevant anymore, and (b) simplify the decision tree here a little bit. This is getting pretty hard to follow.

**Comment by hanhanW** [comment]
Please hold off the review a bit, I have to fix peeling pipeline and update PR description.

**Comment by dcaballe** [comment]
> > Missing a bit of context. What is this trying to achieve? What does `expt` mean?
> > It general, it looks good. My only concern is that we are adding yet another independent pipeline for matmul. I'm worried about the scalability/composition of this. Since you are introducing a "benefit" function, I wonder if we could integrate this new strategy as part of one of the main pipelines, similar to what we did for peeling and use that strategy based on the benefit function. Does that make any sense?
> 
> Second this. I think we need to (a) consolidate different experiments into smaller number of pipelines, and maybe discard things that are not relevant anymore, and (b) simplify the decision tree here a little bit. This is getting pretty hard to follow.

Oh, `expt` means experimental! Sorry :)

Yes, I'm missing some context but in general I would refrain from introducing too specific optimizations and/or "magic numbers" that are aimed at optimizing a single model. I've been there, it's super appealing :) but it can get out of control *extremely fast* and this is something that can really mess up the code generation infra very quickly. We would also be setting non-realistic performance expectations for that model, because when the approach and/or magic number have to be changed to accommodate other requirements/models, the performance will go down.

Not as part of this patch but perhaps we could try to unify the padding and peeling pipelines (peeling is just a flag right now) and also centralize the way to choose the tile sizes. It could be a utility that gathers all the necessary information (pre-vectorization strategy (i.e., padding, peeling, none), target information (vector length, memory hierarchy (in the future), etc.), tensor shapes, etc.) to make a final decision on tile sizes and returns them in a constant container so that they can't be modified anywhere else. That would help with temporary experiments since those changes would be very well localized/contained in a single place.

**Comment by hanhanW** [comment]
> Missing a bit of context. What is this trying to achieve? 

The context about the new strategy is to apply tile interchange around matmul ops. I'm sorry that some context are mixed here. What is going to be needed in my mind is:

1. We'll need to add a level of tiling about L1 caches. In this context, we're going to have triple tiling expert.
2. Apply tile_interchange with fusion cases at L1 tiling level.

In this PR, I only set L1 tiling sizes for reduction dim. The L1 tiling sizes for parallel dims are not considered in the PR. I've done some experiments about it, but could not get perf improvement. So I decide to go with adding another tiling along reduction first, and deal with parallel dims later.

> Oh, expt means experimental! Sorry :)

The intention was creating an experimental pipeline, and experimenting new founds on it. I should have to enable it through flags if it's the intention. Since every model gets improved, we can make it default on. I update the old pipeline and reuse it for new strategy.

 > Not as part of this patch but perhaps we could try to unify the padding and peeling pipelines

Padding is quite tricky because the order matters. IMO, we're able to unify the vectorization strategy for peeling and none. I'd suggest to leave padding and peeling separately for a while, and only keep two pipelines. One is for padding, and the other is for peeling and none. Overall, I totally agree with what you've mentioned. Big +1 on it! :D

> It could be a utility that gathers all the necessary information (pre-vectorization strategy (i.e., padding, peeling, none), target information (vector length, memory hierarchy (in the future), etc.), tensor shapes, etc.) to make a final decision on tile sizes and returns them in a constant container so that they can't be modified anywhere else. That would help with temporary experiments since those changes would be very well localized/contained in a single place.

The decision tree is getting complicated. Maybe it's time to collect data and try to simplify it.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:551]: We should add a comment about why we are using these tile sizes and what is the plan to make this generic. I would strongly prefer if we could avoid adding model-specific hard-coded numbers.

**Review by dcaballe — COMMENTED** [review]
Adding a first round of comments. I'll take another look when we clarify some of the comments. Thanks!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: They are magic numbers actually, was derived from my experiments. We probably have to make them be related to memory hierarchy. I'll add a TODO to it.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:343]: oops, it indicates the number of levels.. Let me replace it with `numLevels`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:360]: All the ops will be vectorized in this case. Because the anchorOpName is empty. In this case, the CodegenStrategy will look into all the Linalg ops and try to vectorize them.

(Actually, I'm surprised that the old code works... CodegenStrategy is a filter-based transform manager, each transform is applied to the ops that was changed/matched from the previous transform. I thought other ops (like linalg.fill, etc) do not have reduction dims. In this case, tiling won't put a filter/marker to the op, and vectorize should just skip them. Anyway, separating them into different CodegenStrategy passes look good to me. We aim to vectorize all the operations.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:350]: It's applicable, but not directly match the for loop. IMO, we can move it to https://github.com/iree-org/iree/blob/2c5ca4963ca22260aebdfa88e1c04b8b04455fbc/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.h#L26-L34 and explain why we have `ParallelTiles` and `ReductionTiles` separately.

(In the future, I'm thinking to add an addtional pass to break a list of tiling sizes into `ParallelTiles` and `ReductionTiles`. We should not consider this kind of limitation when configuring the tiling sizes. At KernelConfig level, we should just determine the tiling sizes for different level of tiling, like distribution, L1 memory, register, etc. And leave the details to later stages. In this context, the comment should be put to the additional pass.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:711]: It's enabled by default because I observed that it improves all our benchmarks on x86. I'm not able to measure RISC-V perf on my side, so I break the condition here. It's might not mature enough to RISC-V.

Do you think that we should keep x86 pipeline and RISC-V pipeline using same strategy and configurations? I'm happy to apply the change to RISC-V as well. The difficulty to me is that I'm not able to measure RISC-V performance, or triage the issues if there are regressions.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:487]: This sounds really good! I'll try to send a prototype for quick review. (and do some cleanup later).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:701]: I don't have a concrete idea yet, but I'll make this part smarter.

**Review by dcaballe — APPROVED** [review]
Ok, I guess we can introduce this to run some experiments but I don't think should go much further than this without refactoring/cleaning this up a bit. I'm sending a patch for padding/peeling that should help a bit on that front.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:778]: Let me take a stab at rebasing on top of your PR. I might can simplify the logic a bit.

**Comment by hanhanW** [comment]
I rebased the PR on top of tree, please take a look! Thanks!

**Review by dcaballe — APPROVED** [review]
Thanks!

**Pull request merged**
